### PR TITLE
feat: fixes and tests for autoreplace support for replacements

### PR DESCRIPTION
**Pull request by t-kulmburg**
As we are also interested in the replace feature, I had a look at the proposed code changes.
To make my changes more clear, I am creating this PR against @JamieMagee's [PR](https://github.com/renovatebot/renovate/pull/13211)
I hope this is also alright with you @JamieMagee.
After fixing building errors I added the name-replacement into the doAutoReplace function as they share most of their logic.
If you have any good ideas for better structure, I can refactor the code further.

I also crated a [gitlab demo-repo](https://gitlab.com/t.kulmburg/replacement-demo/-/blob/main/renovate.json) containing a dependency for all managers listed in the PR of @JamieMagee.
As can be seen in the [repo README](https://gitlab.com/t.kulmburg/replacement-demo/-/blob/main/README.md) and the [Renovate created merge requests](https://gitlab.com/t.kulmburg/replacement-demo/-/merge_requests), more then half of them are already working with my rather minimal changes.
I also added a brief description of the problems the other managers currently have.
To test the feature even further I added unit tests for Managers that worked in the demo-repo.

My question now would be how you would like to proceed with this topic in general.
Getting the feature to work for all supported managers will probably require a lot more work,
but supporting only the currently working ones will require a detailed documentation on the current status.

I am willing to spend some more time on this in the next days/weeks, so reviews would be appreciated.

### CODE REVIEW

**Comment by t-kulmburg** [comment]
I tested this after merging with develop. Currently without merging the tests are working when started separately, but all at once run out of memory.

File: lib/workers/repository/update/branch/auto-replace.spec.ts [hunk 1]
```diff
@@ -200,76 +200,656 @@ describe('workers/repository/update/branch/auto-replace', () => {
...
-      upgrade.newName = 'library/ubuntu';
-      upgrade.packageFile = 'Dockerfile';
-      const res = await doReplacementAutoReplace(
-        upgrade,
-        dockerfile,
-        reuseExistingBranch
-      );
-      expect(res).toBe(dockerfile);
-    });
-
```
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.spec.ts:226]: why this case is removed?

File: lib/workers/repository/update/branch/auto-replace.spec.ts [hunk 2]
```diff
@@ -200,76 +200,656 @@ describe('workers/repository/update/branch/auto-replace', () => {
...
-      const res = await doReplacementAutoReplace(
-        upgrade,
-        hcl,
-        reuseExistingBranch
-      );
+      const res = await doAutoReplace(upgrade, hcl, reuseExistingBranch);
       expect(res).toBe(hcl.replace(upgrade.depName, upgrade.newName));
     });
 
-    it('fails with old name in depname', async () => {
```
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.spec.ts:243]: does this no longer fail?

File: lib/workers/repository/update/branch/auto-replace.spec.ts [hunk 3]
```diff
@@ -200,76 +200,656 @@ describe('workers/repository/update/branch/auto-replace', () => {
...
   });
+
+  describe('doReplacementAutoReplaceNewDepName', () => {
+    let reuseExistingBranch: boolean;
+    let upgrade: BranchUpgradeConfig;
+
+    beforeEach(() => {
+      upgrade = {
+        ...JSON.parse(JSON.stringify(defaultConfig)),
+      };
```
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.spec.ts:244]:
```suggestion
      upgrade = getConfig();
```

`getConfig` will always return a fresh config

**Review by viceice — CHANGES_REQUESTED** [review]
Please start a fresh PR and target main instead.

Don't do refactorings in this feature PR. extract required refactorings to a seperate PR, which should merged before this feature.

**Comment by viceice** [comment]
> I tested this after merging with develop. Currently without merging the tests are working when started separately, but all at once run out of memory.

you should start fresh from main, then the memory issue should be gone

**Comment by t-kulmburg** [comment]
Ok so just that I understand you correctly, you just want me to create a new PR with this changes to main?
(and after considering your comments of course)

Just because I started from @JamieMagee 's PR and refactored/fixed his work, and I did not want to just create one and make it seem like I did this all on my own.

**Comment by viceice** [comment]
> Ok so just that I understand you correctly, you just want me to create a new PR with this changes to main? (and after considering your comments of course)
> 
> Just because I started from @JamieMagee 's PR and refactored/fixed his work, and I did not want to just create one and make it seem like I did this all on my own.

no problem @JamieMagee is very busy, so i 'm pretty sure he's happy when somebody finish his work.

**Comment by JamieMagee** [comment]
@t-kulmburg thanks for opening this. It looks like you opened #17476 targeting `main`, instead of my branch. Does that mean we can close this one?

**Pull request closed without merging**